src/System.IO.IsolatedStorage/src/System/IO/IsolatedStorage/IsolatedStorage.cs

```diff
@@ -0,0 +1,194 @@
+            }
+        }
+
+        [CLSCompliant(false)]
+        [Obsolete("IsolatedStorage.MaximumSize has been deprecated because it is not CLS Compliant.  To get the maximum size use IsolatedStorage.Quota")]
+        public virtual ulong MaximumSize
+        {
+            get
+            {
+                if (_validQuota)
+                    return _quota;
+
+                throw new InvalidOperationException(string.Format(SR.IsolatedStorage_QuotaIsUndefined, nameof(MaximumSize)));
+            }
+        }
+
+        public virtual long Quota
+        {
+            get
+            {
+                if (_validQuota)
+                    return (long)_quota;
+
+                throw new InvalidOperationException(string.Format(SR.IsolatedStorage_QuotaIsUndefined, nameof(Quota)));
+            }
+
+            internal set
+            {
+                _quota = (ulong)value;
+                _validQuota = true;
+            }
+        }
+
+        public IsolatedStorageScope Scope
+        {
+            get; protected set;
+        }
+
+        protected virtual char SeparatorExternal
+        {
+            get { return Path.DirectorySeparatorChar; }
+        }
+
+        protected virtual char SeparatorInternal
+        {
+            get { return '.'; }
+        }
+
+        public virtual bool IncreaseQuotaTo(long newQuotaSize)
+        {
+            return false;
+        }
+
+        public abstract void Remove();
+
+        protected string IdentityHash
+        {
+            get; private set;
+        }
+
+        protected void InitStore(IsolatedStorageScope scope, Type appEvidenceType)
+        {
+            InitStore(scope, null, appEvidenceType);
+        }
+
+        protected void InitStore(IsolatedStorageScope scope, Type domainEvidenceType, Type assemblyEvidenceType)
+        {
+            Scope = scope;
+
+            // NetFX (desktop CLR) IsolatedStorage uses identity from System.Security.Policy.Evidence to build
+            // the folder structure on disk. It would use the "best" available evidence in this order:
+            //
+            //  1. Publisher (Authenticode)
+            //  2. StrongName
+            //  3. Url (CodeBase)
+            //  4. Site
+            //  5. Zone
+            //
+            // For CoreFx StrongName and Url are the only relevant types. By default evidence for the Domain comes
+            // from the Assembly which comes from the EntryAssembly(). We'll emulate the legacy default behavior
+            // by pulling directly from EntryAssembly.
+
+            Assembly assembly = Assembly.GetEntryAssembly();
+            AssemblyName assemblyName = assembly.GetName();
+            Uri codeBase = new Uri(assembly.CodeBase);
+            object identity = null;
+
+            string hash = Helper.GetNormalizedStrongNameHash(assemblyName);
+            if (hash != null)
+            {
+                hash = $"StrongName{SeparatorInternal}{hash}";
```

Comment:
I just [confirmed](http://tryroslyn.azurewebsites.net/#f:>il/A4VwRgNglgxgBDCBDAziuAVApigLgKAG985S5gAnAe1yxloBM4A3KC3EJCBACyQrgBlLMH5JcVCgEkAdrQoyucQnADmWXMri4e1AO5wZICBADccAL6WSZG6VCRYLKlCYAxKlQAUeClBmqcHwoPACUynZkpME8cAC8cABEgrjUAQBySAC2WIlwANRCImIS0nJYCkqFMaaRFvgWQA=), the following:

```c#
public class Test
{
    protected virtual char SeparatorInternal { get { throw null; } }
    
    public void Foo(string hash) {
        hash = "StrongName" + SeparatorInternal + hash;
    }
}
```

Compiles to:

```
    .method public hidebysig 
        instance void Foo (
            string hash
        ) cil managed 
    {
        // Method begins at RVA 0x2054
        // Code size 29 (0x1d)
        .maxstack 3
        .locals init (
            [0] char
        )

        IL_0000: nop
        IL_0001: ldstr "StrongName"
        IL_0006: ldarg.0
        IL_0007: callvirt instance char Test::get_SeparatorInternal()
        IL_000c: stloc.0
        IL_000d: ldloca.s 0
        IL_000f: call instance string [mscorlib]System.Char::ToString()
        IL_0014: ldarg.1
        IL_0015: call string [mscorlib]System.String::Concat(string, string, string)
        IL_001a: starg.s hash
        IL_001c: ret
    } // end of method Test::Foo
```

Also, approximate micro benchmark:

```c#
using System;
using System.Diagnostics;

public class Program
{
    public static void Main()
    {
        TimeAction("Old", () => Old("Foo"));
        TimeAction("New", () => New("Foo"));
    }

    private static char SeparatorInternal => 'a';

    private static string Old(string hash) => $"StrongName{SeparatorInternal}{hash}";

    private static string New(string hash) => "StrongName" + SeparatorInternal + hash;

    private static void TimeAction(string prefix, Action action, int times = 5, int iterations = 10000000)
    {
        var sw = new Stopwatch();
        for (int i = 0; i < times; i++)
        {
            int gen0 = GC.CollectionCount(0);
            sw.Restart();
            for (int iter = 0; iter < iterations; iter++)
            {
                action();
            }
            sw.Stop();
            Console.WriteLine($"{prefix}: Time: {sw.Elapsed.TotalSeconds}\tGC0: {GC.CollectionCount(0) - gen0}");
        }
    }
}
```

Results (Windows 10 x64):

```
Old: Time: 1.3506488    GC0: 267
Old: Time: 1.3122971    GC0: 267
Old: Time: 1.294516     GC0: 267
Old: Time: 1.3067892    GC0: 267
Old: Time: 1.2916698    GC0: 267
New: Time: 0.2853555    GC0: 209
New: Time: 0.2842609    GC0: 210
New: Time: 0.2908145    GC0: 210
New: Time: 0.2880131    GC0: 210
New: Time: 0.2878425    GC0: 210
```